File: spring/src/test/java/cardsproject/controller/marketplace/CouponControllerTest.java
package cardsproject.controller.marketplace;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class CouponControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/coupons"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/coupons")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"code\": \"test\", \"discountValue\": 0.01, \"validFrom\": \"2024-01-01T00:00:00\", \"validUntil\": \"2024-01-01T00:00:01\", \"maxUses\": null }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/coupons/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/coupons/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404 || status == 500 || status == 501;
            });
    }
    @Test
    void create_fails_when_discount_value_positive_violated() throws Exception {
        // Discount value must be greater than zero → 400 (Bean Validation)
        mockMvc.perform(post("/api/coupons")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"code\": \"test\", \"minOrderValue\": 0.00, \"usesCount\": 1, \"validFrom\": \"2024-01-01T00:00:00\", \"validUntil\": \"2024-01-01T00:00:00\", \"isActive\": true, \"discountType\": \"PERCENT\", \"maxUses\": 1, \"discountValue\": 0.00 }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_percent_discount_range_violated() throws Exception {
        // Percent discount must be between 1 and 100: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/coupons")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"code\": \"test\", \"minOrderValue\": 0.00, \"usesCount\": 1, \"validFrom\": \"2024-01-01T00:00:00\", \"validUntil\": \"2024-01-01T00:00:00\", \"isActive\": true, \"discountType\": \"PERCENT\", \"discountValue\": 101 }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_uses_not_exceed_max_violated() throws Exception {
        // Coupon uses count cannot exceed max_uses: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/coupons")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"code\": \"test\", \"discountType\": \"PERCENT\", \"discountValue\": 0.00, \"minOrderValue\": 0.00, \"usesCount\": 1, \"validFrom\": \"2024-01-01T00:00:00\", \"validUntil\": \"2024-01-01T00:00:00\", \"isActive\": true, \"maxUses\": 1 }"))
            .andExpect(status().isBadRequest());
    }
}

File: phoenix/test/marketplace/coupon_controller_test.exs
defmodule CardsProjectWeb.Marketplace.CouponControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "code" => "test",
    "discount_value" => 1,
    "min_order_value" => "0.00",
    "uses_count" => 0,
    "valid_from" => ~N[2024-01-01 00:00:00],
    "valid_until" => ~N[2024-12-31 23:59:59],
    "is_active" => true,
    "discount_type" => "Percent"
  }

  describe "GET /api/coupons" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/coupons")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "GET /api/coupons?q=test" do
    test "returns 200 with search results", %{conn: conn} do
      conn = get(conn, "/api/coupons?q=test")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/coupons" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/coupons", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
    test "fails when valid_until_after_valid_from violated", %{conn: conn} do
      params = Map.put(@valid_params, "valid_until", 0)
      conn = post(conn, "/api/coupons", params)
      assert conn.status in [400, 422]
    end
    test "fails when discount_value_positive violated", %{conn: conn} do
      params = Map.put(@valid_params, "discount_value", 0)
      conn = post(conn, "/api/coupons", params)
      assert conn.status in [400, 422]
    end
  end

  describe "GET /api/coupons/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_coupon(@valid_params)
      conn = get(conn, "/api/coupons/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/coupons/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_coupon(@valid_params)
      conn = put(conn, "/api/coupons/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

end
